### ISSUE TITLE
Can't generate boarding pass

### ISSUE THREAD

**Issue by ianbale** [alexandercerutti/passkit-generator#25]
Hi,

I've hit a problem trying to generate a boarding pass. 

I have managed to generate a sample event type pass with no problem at all. It's just the boarding pass that is giving me problems.

I'm seeing this error:

`  Schema validation failed due to error: "transitType" is not allowed +0ms`

I do have transitType in the boadingPass object as you can see in my JSON

```
{
  "formatVersion": 1,
  "organizationName": "Hazardous Frog LTD",
  "passTypeIdentifier": "pass.com.hazardousfrog.xxxxx",
  "serialNumber": "placeholder",
  "teamIdentifier": "AAAAAAAAAAA",
  "description": "HF Rail Ticket",
  "relevantDate": "2019-12-08T13:00-08:00",
  "locations": [
    {
      "longitude": -122.3748889,
      "latitude": 37.6189722
    },
    {
      "longitude": -122.03118,
      "latitude": 37.33182
    }
  ],
  "barcodes": [{
    "message": "06FB3SHXG3Y00R1JTETQMAUZYHXLALMCRDZPWTFPFWKKBDSZSQEQTVYWSZZWWWNXSSUAXCDJDQTZMGODMFIRJCEWEKACPBNOUGZCDRESKGTKFTZNYDURPFILJFCVAHTQFHELJCXPYBCJJIIJPIRCVECCLQWFMGROBHRGFIRKPNJVKOOYCXTHRAIFSFXQQUVCARZROCACWBEGEWUVFKTDKNSQCACSGWUDBVDHMJMGH",
    "format": "PKBarcodeFormatAztec",
    "messageEncoding": "iso-8859-1"
  }],
  "logoText": "...Logo...",
  "foregroundColor": "rgb(0, 0, 0)",
  "backgroundColor": "rgb(255, 248, 242)",
  "boardingPass": {
    "transitType": "PKTransitTypeTrain",
    "headerFields": [
      {
        "label": "journey",
        "key": "journey",
        "value": "Field E0001",
        "textAlignment": "PKTextAlignmentLeft"
      }
    ],
    "primaryFields": [
      {
        "key": "depart",
        "label": "depart",
        "value": "27/11/2019",
        "textAlignment": "PKTextAlignmentLeft"
      },
      {
        "key": "arrive",
        "label": "arrive",
        "value": "27/11/2019",
        "textAlignment": "PKTextAlignmentRight"
      }
    ],
    "secondaryFields": [
      {
        "key": "status",
        "label": "status",
        "value": "ADULT",
        "textAlignment": "PKTextAlignmentLeft"
      },
      {
        "key": "validuntil",
        "label": "valid until",
        "value": "27/11/2019",
        "textAlignment": "PKTextAlignmentRight"
      }
    ],
    "auxiliaryFields": [
      {
        "key": "tickettype",
        "label": "ticket type",
        "value": "SOS",
        "textAlignment": "PKTextAlignmentLeft"
      },
      {
        "key": "route",
        "label": "route",
        "value": "Any Permitted",
        "textAlignment": "PKTextAlignmentRight"
      }
    ],
    "backFields": [
      {
        "key": "itinerary",
        "label": "itinerary",
        "value": "itinerary goes here..."
      },
      {
        "key": "freeinformation",
        "value": "Field E0044"
      },
      {
        "key": "contactinformation",
        "label": "Contact Details",
        "value": "Field E0022"
      }
    ]
  }
}
```

if I omit transitTyoe (as a test) I get this error instead:

`UnhandledPromiseRejectionWarning: Error: Error: Cannot proceed with pass creation. transitType field is required for boardingPasses.`

I'm using v 1.6.4 with Node 8.10.0

Do you have any idea what I am doing wrong here?

**Comment by ianbale** [comment]
I've done a bit of hacking about. As far as I can see there are two issues:

1) The JOI schema does not permit transitType in passDict in schema.js
adding transitType:Joi.string() resolves that

but then we hit another error:

2) transitType is not copied into **this** scope in _patch (pass.js) and so throws  error:

` UnhandledPromiseRejectionWarning: Error: Error: Cannot proceed with pass creation. transitType field is required for boardingPasses.`

Making this modification to copy transitType into scope if the type is boardingPass

```
if (this.type === "boardingPass") {                    
  if (passFile.boardingPass.transitType)
  {
    this.transitType = passFile.boardingPass.transitType
  }
  else
  {
    throw new Error(formatMessage("TRSTYPE_REQUIRED"));
  }               
}
```

does allow the pass to be generated...

Probably not the best place to do that, but at least its a start in terms of location the source of the issue...

**Comment by alexandercerutti** [comment]
Hi @ianbale,
it is weird that it does not save it in your pass.json: if you look at line [671](https://github.com/alexandercerutti/passkit-generator/blob/master/src/pass.js#L671), in `_patch()`, that's exactly what I'm already doing.

TransitType is not a property to be passed in `overrides`, but one of the Fields, like `.primaryFields`.
So you have to do `(new Pass({ ... }).transitType = "..."`. Inside the code, a setter is available and will validate your type against a schema.

Let me know! 😊

**Comment by ianbale** [comment]
I'm wondering if I have a misunderstanding....?

I've created the entire ticket data in my JSON. I have the basic fields, plus the boardingPass object containing all the fields, including the transitType. I've not used the API to pass in any additional data.

That seems to work for everything except the transitType. That is the only field (that I am aware of) that is not brought into your **this** scope when parsing the JSON. Hence, it throws an error prior to line 671.

```
		if (this.type === "boardingPass" && !this.transitType) {
			throw new Error(formatMessage("TRSTYPE_REQUIRED"));
		}

		passFile[this.type]["transitType"] = this.transitType;
```

In fact, as I pointed out, your JOI does not permit that to be in the JSON - although the remainder of the boardingPass object can be.

So..I guess if the intention is that we do not include the boardingPass part in the JSON, but rather attach that using the API, then all is well (although that would be a shame since it almost works already). But, if putting all the data in the JSON should work too, then the changes I've outlined would be needed - although, something a little less rough and ready than I threw together to debug the issue!

BTW: I'm going to be using this to generate UK train tickets. Will send you the detail once it's live (should be October).

**Comment by alexandercerutti** [comment]
No okay, you are right. I read both messages better and found what you mean. I also tested and got about the same result as yours.
As I was thinking, this is a problem related to the fact that the current version of passkit-generator, does not fully parse and save the pass.json. It will be implemented on v2 (I'm already working on it, and it is beautiful 😂).

Do you mind to remove it from the pass.json and set it as I said in my previous message, until the new version?

Also, what I'm thinking, is that I designed this package to be model-specific. I mean, you have a specific model for a specific type and sort of things. What I'm doing right now, is to let `transitType` to be defined in runtime, while I should only validate it once in pass. What do you think about?

**Comment by ianbale** [comment]
From the perspective of this application I am going to have 4 ticket templates (for four different classes of tickets). They will all be train transit types. So being able to specify that at runtime is not something I, personally, would need to do, but I suppose other people might well want to do that.

Without giving it a lot of thought, just off the top of my head... What, at first glance at least, seems like a nice easy way to do this would be to have an option to define the entire ticket in the JSON file. All the fields defined but none of the values.

Then we can simply create a set of key / value pairs to provide the data. The generate function simply combines the two to produce a data file with the structure and the data. No real need to have the API then...

JSON (defines the fields / positions and possibly labels - although in my case, the labels are also dynamic)
```
    "primaryFields": [
      {
        "key": "origin",
        "textAlignment": "PKTextAlignmentLeft"
      },
      {
        "key": "destination",
        "textAlignment": "PKTextAlignmentRight"
      }
    ],
```

Key value pairs to populate the above (and main attribute values):
```
{
  serialNumber : "WZYZ0001",
  groupingIdentifier : "ABCD1234",
  description: "Rail Ticket EUS-MAN",
  origin : {label:"LONDON EUSTON", value:"EUS"}
  destination : {label:"MANCHESTER PICCADILLY", value:"MAN"}
{
```

I'll modify my test code tomorrow to create my ticket via the API rather than the JSON. As soon as I have that working, I'll crack on with integrating it into my application along with the PDF ticket generation (which is already complete).

What's your ETA for V2?

**Comment by alexandercerutti** [comment]
I'm sorry, I'm not understanding what you mean with

> a nice easy way to do this would be to have an option to define the entire ticket in the JSON file. All the fields defined but none of the values.

Define an entire ticket?
Anyway, look at the API documentation. You can use all the `Array.prototype` methods for all the Pass Fields (`primaryFields`, `secondaryFields`, ...). So you can push and pop, for example, the fields you need.

About the ETA, I don't have one yet as I develop it mainly on the weekends or the evenings.
Currently, it is in early alpha stage: it doesn't have tests yet, nor is working. But it will include a radically changed API. You can look at the `Project` tab above to see some of the features it will include. Once it will start being stable, I'll publish it in a branch to let users testing. 😊

**Comment by ianbale** [comment]
This is what I had in mind...

This would be my JSON file:

```
{
  "formatVersion": 1,
  "organizationName": "Hazardous Frog LTD",
  "passTypeIdentifier": "pass.com.hazardousfrog.rail",
  "serialNumber": "placeholder",
  "teamIdentifier": "54GQCP5JCQ",
  "description": "placeholder",
  "relevantDate": "placeholder",
  "locations": [],
  "groupingIdentifier" : "placeholder",
  "logoText": "placeholder",
  "foregroundColor" : "rgb(0, 0, 0)",
  "backgroundColor" : "placeholder",
  "labelColor" : "rgb(121, 121, 121)",
  "boardingPass": {
    "transitType": "PKTransitTypeTrain",
    "headerFields": [
      {
        "key": "journey",
        "textAlignment": "PKTextAlignmentLeft"
      }
    ],
    "primaryFields": [
      {
        "key": "origin",
        "textAlignment": "PKTextAlignmentLeft"
      },
      {
        "key": "destination",
        "textAlignment": "PKTextAlignmentRight"
      }
    ],
    "secondaryFields": [
      {
        "key": "status",
        "textAlignment": "PKTextAlignmentLeft"
      },
      {
        "key": "validuntil",
        "textAlignment": "PKTextAlignmentRight"
      }
    ],
    "auxiliaryFields": [
      {
        "key": "tickettype",
        "textAlignment": "PKTextAlignmentLeft"
      },
      {
        "key": "route",
        "textAlignment": "PKTextAlignmentRight"
      }
    ],
    "backFields": [
      {
        "key": "itinerary",
      },
      {
        "key": "freeinformation",
      },
      {
        "key": "contactinformation",
      }
    ]
  }
}
```

and my values:

```
let values = {
  
  serialNumber : "ABCD1234",
  description : "Rail ticket EUS-MAN",
  relevantDate : "2019-12-08T13:00-08:00",
  locations : [
    {
      "longitude": -122.3748889,
      "latitude": 37.6189722
    },
    {
      "longitude": -122.03118,
      "latitude": 37.33182
    }
  ],
  groupingIdentifier : "XYZ12345",
  logoText : "27 Jun 2019",
  backgroundColor : "rgb(255, 248, 242)",

  journey : {label:"",value:"EUS-MAN"},
  origin : {label:"LONDON EUSTON",value:"EUS"},
  destination : {label:"MANCHESTER PICCADILLY",value:"MAN"},
  status : {label:"ADULT/CHILD",value:"Child"},
  validuntil : {label:"VALID AT",value:"11:20"},
  tickettype : {label:"TICKET TYPE",value:"Advance Single"},
  route : {label:"ROUTE",value:"Virtin Trains Only"},
  itinerary : {label:"ITINERARY",value:"long text goes here..."},
  freeinformation : {label:"",value:"long text goes here..."},
  contactinformation : {label:"",value:"long text goes here..."}
}
```

then I'd have 

```
let examplePass = new Pass({
    model: "./eticket",
    certificates: {
        wwdr: "./certs/AppleWWDRCA.pem",
        signerCert: "./certs/signerCert.pem",
        signerKey: {
            keyFile: "./certs/signerKey.pem",
            passphrase: "secret!"
        }
    }
})

let stream = await examplePass.generate(values)
```
The overrides would not be necessary since VALUES is all override values...

My entire ticked structure is defined in the model, All I need to do outside of that is pass in a set of key/value pairs with the data needed to complete the pass.

I'm a great believer in "simpler is better"...

I'm sure you are aware that the following values are not possible to overrides using the existing overrides feature. The error says something about wrong type for an overrides. Looks like it's strings only?

relevantDate, locations, foregroundColor

All of those are things I need to change on a pass by pass basis, so as it stands, I'd have to generate the JSON on the fly for each pass.

It would also be nice if logo could be overridden as well. My project doesn't actually need that, but I do need to put different graphics in elsewhere (I'm generating tickets for more than one customer. Most of the ticket will be identical, apart from one section that details the retailer.). So I'd imagine someone else will have a use case to have two identical tickets, but with different logos...


Anyway, that's just a couple of suggestions to mull over. What you've doen so far has enabled me to get a wallet pass created quickly and easily without having to wade through all the Apple docs. Much appreciated.

I don't want any details published until after this goes live - hence there is nothing about this on our website yet (new website built, and ready to replace the current once as soon as we go live). I've had too many instances of innovative ideas with UK train retailing quickly appropriated by our competitors. Lots of unique things in this latest project that have to be kept under wraps until its launched. But once it is live I'll give you a link and overview of the project. Feel free to use it to show off your work... 

Anyway...time to crack on and redo this prototype using your API then once its exactly as I want work out how to integrate it into the main project. The pain bit is going to be delivering a stack of these at once. We could easily be generating 20 or more tickets for a single booking. Not easy to make it all obvious to the customer which once they need when... Much easier with a PDF for each passenger where I can put multiple tickets all in the correct order. I guess no one considered a user case where multiple passes are needed for a single journey...

**Comment by ianbale** [comment]
Sorry to be a pain, but I've just hit another problem.

I'm creating multiple passes in parallel:

```
makeOnePass = async (passConfig,passData) => {

  let pass = new Pass(passConfig)

  pass.barcode(passData.barcode)
  pass.expiration(passData.expiry)

  for (let key in passData.relevance)
  {
    pass.relevance(key,passData.relevance[key])
  }

  for (let key in passData.ticketData)
  {
    let data = passData.ticketData[key]

    for (let i=0;i<data.length;i++)
    {
      pass[key].push(data[i])
    }
  }

  pass.transitType = "PKTransitTypeTrain"

console.log(passConfig.overrides.serialNumber,JSON.stringify(pass.primaryFields,null,2))

  let writestream = await fs.createWriteStream(`./tickets/${passConfig.overrides.serialNumber}.pkpass`)

  // Generate the stream, which gets returned through a Promise
  let stream = await pass.generate()
  stream.pipe(writestream)
}
```

and calling that like this:

```
  for (let s=0;s<serialNumbers.length;s++)
  {
    let serialNumber = serialNumbers[s]
    let departDate = "30 Jun 2019"
    let journey = "EUS-XXX"

    let overrides = {
      serialNumber: serialNumber,
      groupingIdentifier : groupingIdentifier,
      logoText : departDate,
      description : "Rail Ticket " + journey,
    }

    let passConfig = {...eTicketPassConfig,model:"./eticket",overrides:overrides}

    let passData = {
      barcode : {
        "message": "06FB3SHXG3Y00R1JTETQMAUZYHXLALMCRDZPWTFPFWKKBDSZSQEQTVYWSZZWWWNXSSUAXCDJDQTZMGODMFIRJCEWEKACPBNOUGZCDRESKGTKFTZNYDURPFILJFCVAHTQFHELJCXPYBCJJIIJPIRCVECCLQWFMGROBHRGFIRKPNJVKOOYCXTHRAIFSFXQQUVCARZROCACWBEGEWUVFKTDKNSQCACSGWUDBVDHMJMGH",
        "format": "PKBarcodeFormatAztec",
        "messageEncoding": "iso-8859-1",
        "altText" : serialNumber
      },
      expiry : new Date(),
      relevance : {
        location : [
          {
            "longitude": -122.3748889,
            "latitude": 37.6189722
          },
          {
            "longitude": -122.03118,
            "latitude": 37.33182
          }
        ],
        relevantDate : new Date()
      },
      ticketData : {
        headerFields : [
          {
            "label": "",
            "key": "journey",
            "value": "EUS-MAN",
            "textAlignment": "PKTextAlignmentLeft"
          }
        ],
        primaryFields: [
          {
            "key": "depart",
            "label": "LONDON EUSTON",
            "value": "EUS",
            "textAlignment": "PKTextAlignmentLeft"
          },

     .... etc.

    }

    promises.push(makeOnePass(passConfig,passData))
  }

  await Promise.all(promises)
```

As a test I'm passing identical data in for each - apart from the serialNumber which is unique.

So I should end up with 3 almost identical tickets. However, what I end up with is one correct ticket and 2 blank ones.

Using some debugging I can confirm that I am definitely pushing the relevant headerFields etc. data into pass. But, if I dump the content of pass.headerFields to the console immediately afterwards, I see it's only populated for the first ticket. 

It appears that although I am creating three pass instances with 'let pass = new Pass(passConfig)', they are not actually unique instances. It looks like I am working with a single instance with shared data?

If instead of doing them in parallel, I do them sequentially, then it works fine.

**Comment by ianbale** [comment]
Re the above issue:

The problem is in fieldsArray.js

You've defined fieldsKeys outside of the class, so it is shared across all instances.

Change the constructor to:

```
constructor(...items) {
    super(...items);

    this.fieldsKeys = new Set();
 }
```

emptyUnique to :

```
 emptyUnique() {
    this.fieldsKeys.clear();
 }
```

and prefix all other references to emptyUnique with this.

And finally, in pass.js replace:

```
FieldsArray.emptyUnique();
```

with
```
this._fields.forEach(a => this[a].emptyUnique());
```

That sorts that problem out.

I guess I should really do an update and send you a pull request....
Will see if I can sort that later.

**Comment by alexandercerutti** [comment]
One message per time 😂

> I'm sure you are aware that the following values are not possible to overrides using the existing overrides feature. The error says something about wrong type for an overrides. Looks like it's strings only?

Which values, exactly? All of them? Or the one below? `relevance`, `locations`, `foregroundColor`, `logoText`.

If you mean these four, for the first two, you can use `.relevance()` method (it will be also changed in v2). You can pass `foregroundColor` in Overrides, you cannot override `logoText` as package's architecture choice, but we may discuss about it. If you are interested, you can comment in [this issue - #21](https://github.com/alexandercerutti/passkit-generator/issues/21) that I left open in case other people would need it.


Whoa, 20 tickets per user?! Really? I think this is not suggested at all, for a matter of User Experience!
Usually Apple allow it to generate multiple and link them together, but usually you (user) can get up to 3 or maximum 4 passes all together. Sadly there is no way to generate passes in bulk. I mean, you cannot let Wallet add passes in bulk, but it would be awesome.

About the second issue, I'll reply you in the new issue topic asap. 😊

About your project, once you'll release it, feel free to open an issue topic to tell me that 😉. I'm really glad you are using my package 😄

**Comment by ianbale** [comment]
20 tickets is small... For physical tickets being collected from a machine we have a hard limit of 99 tickets. We've managed to hit that limit quite frequently... Not great usability, but it does save the customer money which they do seem to like ;-)

**Comment by alexandercerutti** [comment]
It is small amount of numbers, but think for a user to tap 20 times on their screen (with an interval of about 4 or 5 second between one ticket and another). It might be annoying. You should reduce this as much as possible imho 😉

**Comment by ianbale** [comment]
Welcome to the wonderful world of rail tickets in the UK! THE most complex transport ticketing system anywhere in the world...

**Comment by alexandercerutti** [comment]
Beautiful 😂

**Comment by gweit** [comment]
@ianbale excuse me sir, are you sending up to 20 (or more) Tickets to the same device? May it be an option to send only one and update it accordingly?

**Comment by ianbale** [comment]
@gweit 

It has to be multiple passes...since we are selling multiple tickets. There is a limit on what can be updated remotely, so even if we could rely upon 100% network coverage to be able to send one ticket and swap it out for the next one when they customer gets to the end of their journey, we'd not be able to change all the necessary fields. I can think of a half dozen other showstoppers for that off the top of my head.

In pretty much any other form of transport you'd have a maximum of 2 tickets - outbound and inbound.

With UK rail the pricing is so messed up that it is often cheaper to buy multiple tickets for each direction. So rather than buying a ticket from Bristol to London, you might buy one from Bristol to Bath Spa, one from Bath Spa to Swindon, another from Swindon to Reading and one from reading to London... And same again for the way back...

All of those ought to be grouped together for convenience.

The savings can be considerable (I've seen examples of 75% saving), so people are using this - just not with wallet tickets - yet...

It could be that the customer is purchasing tickets for several journeys at the same time. So we might need to do that several times over (Although we'd group by journey, so they'd not all appear in the same set).

And a final complication is that you can buy tickets for more than one passenger at a time. So we might be adding passes for several passengers as well as several journeys each with several tickets! Not everyone uses an app, so these passes have to be delivered to the customer via email - yeah, I know, not fun to add lots of passes...

It's a nightmare, but it's not going to get sorted anytime soon. For now I have no option but to live with it and be prepared to be sending large bundles of passes to the customer. But thankfully, that's someone else's problem, but project is simply creating the tickets for a retailer. They have to do the delivery to the customer.

It would be really nice if Apple had a format where we could deliver all the passes for a group as a single file - and specify the order they should be put n within that group. But then Apple didn't look at UK rail as a use case when they designed the wallets! 

Same issue with ITSO (smart cards). Apple's implementation of ITSO on mobile is not compatible with UK rail gates.

Sadly, ATOC are a dumb bunch who have only a slim grasp of the technologies they are specifying. The whole lot needs a clean slate rewrite!

**Comment by alexandercerutti** [comment]
What are ITSO and ATOC?

Anyway, if you group all the passes for the same `groupIdentifier` (overrides), you'll be able to see them all together. You should be ordered for relevance time if I don't remember wrongly.

What might be good for Apple to include, would be (as I said before and as you said too) pass bulking.

**Comment by ianbale** [comment]
My tests show that ordering is the reverse of the time you add to the wallet - so most recently added at the front - not relevant time. Not that that would help since tickets are generally all valid all day. 

ITSO is a standard for smart card tickets (NFC)
ATOC is a front for all the UK train operating companies. The employ people who do not understand the technology to write standards we have to abode by. Very frustrating!

**Comment by alexandercerutti** [comment]
Do you mean in grouped tickets or wallet?

**Comment by ianbale** [comment]
The ordering within the group...

If they have different relevance dates then they show in relevance date order, no matter what order they are added in.

However, If the relevance date is the same (which, in my case, may well be the case for several tickets)
If I add 5 passes in this order : 3,1,4,2,0 they show as 0,2,4,1,3 - reverse order that I added them. That is if I add them directly on my phone from an email. If I add them from my iMac using iCloud sync then the ordering is quite random.

I've not done any tests regarding ordering within the wallet yet...only within a group.

**Comment by alexandercerutti** [comment]
This is weird... 🤔 Don't know how to help you with this.

**Comment by ianbale** [comment]
I'm going to set the relevant date down to the minute for each anticipated leg of the journey. That should force them into the correct order.  Really thought, they are all [usually] valid all day, so its not really correct, but it will work.

**Comment by alexandercerutti** [comment]
So, @ianbale, has been this solved, isn't it?

**Comment by ianbale** [comment]
Yup!

Did you make the change to permit override of logoText?

**Comment by alexandercerutti** [comment]
@ianbale Yes, try it yourself and let me know 😉

**Comment by ianbale** [comment]
All looks god this end :-)

**Comment by alexandercerutti** [comment]
I'm glad of this 😃
